### PR TITLE
feat: add /org/[orgid] facet in pro platform URL

### LINKED ISSUE #9685: Add /org/[organisation name] facet directly in URL on the producers platform
Today we have an hidden facet on the producers platform for the organization name. It would be much better to make that facet directly visible in the URL (like other facets).

This would allow:
- org admins to easily access different organizations
- producers to have access to several organizations (e.g. for sub brands)
- sharing URLs that specify the organization

### PR DESCRIPTION
**Pull request by 4nt0ineB**
### What

Add a visible facet of the current org in the URL for connected users on the pro platform.
Allows admins and pro moderators to easily access pages of multiple orgs and share user-friendly URLs:
 - /org/[orgid]/product/[productid]
 - /org/[orgid]/(nutrition-grades | nova-groups | .. )
- /org/[orgid]/state/en:to-be-exported ...

Eventually, it might helps a new feature that will give producers access to several organizations (sub brands).

This feature is added on top of the existing routing and **does not break old URLs**.

- Refactor of Routing.pm 

### Related issue(s) and discussion

- Fixes #9685

### CODE REVIEW

File: lib/ProductOpener/Display.pm [hunk 1]
```diff
@@ -5321,13 +5333,12 @@ sub search_and_display_products ($request_ref, $query_ref, $sort_by, $limit, $pa
...
 				my $url_path = product_url($product_ref);
-				$product_ref->{url} = $formatted_subdomain . $url_path;
-
+				$product_ref->{url} = $formatted_subdomain . get_org_id_pretty_path() . $url_path;
 				# Compute HTML to display the small front image, currently embedded in the HTML of web queries
 				if (not $api) {
 					$product_ref->{image_front_small_html} = display_image_thumb($product_ref, 'front');
 
-					# For web queries with personal search, we can compute some generated fields we need
+					# For web queries with personal search, we can compute some generated fields we ne<ed
```
Comment (stephanegigandet) [inline lib/ProductOpener/Display.pm:5341]:
```suggestion
					# For web queries with personal search, we can compute some generated fields we need
```

**Comment by stephanegigandet** [comment]
Wow, that's a big refactor! It's a good idea to split in multiple functions.

File: lib/ProductOpener/Routing.pm [hunk 2]
```diff
@@ -192,196 +123,35 @@ Sometimes we modify request parameters (param) to correspond to request_ref:
...
-				xml => single_param("xml")
-			}
-		) if $log->is_debug();
-	}
+	$log->debug("route", {components => \@components,}) if $log->is_debug();
 
-	# /search search endpoint, parameters will be parsed by CGI.pm param()
-	elsif ($components[0] eq "search") {
-		$request_ref->{search} = 1;
+	if (keys %routes == 0) {
```
Comment (stephanegigandet) [inline lib/ProductOpener/Routing.pm:136]: We should probably call load_routes() once in LoadData.pm module instead

**Comment by 4nt0ineB** [comment]
> Wow, that's a big refactor! It's a good idea to split in multiple functions.
:sweat_smile: yes. Basically, I needed to potentially have access to all other paths starting with /org/[orgid] while allowing the public platform to not have that prefix. So I had to brake the path, taking the part I needed (orgid) and with the remaining let the existing routing act normally.

File: lib/ProductOpener/Routing.pm [hunk 3]
```diff
@@ -63,137 +64,561 @@ use Log::Any qw($log);
...
+	{
+		$api_action = 'product';
+	}
+
+	# some API actions have an associated object
+	if ($api_action eq "product") {    # api/v3/product/[code]
+		param("code", $components[3]);
+		$request_ref->{code} = $components[3];
+	}
+	elsif ($api_action eq "tag") {    # api/v3/[tag]/[type]/[tagid]
```
Comment (stephanegigandet) [inline lib/ProductOpener/Routing.pm:305]:
```suggestion
	elsif ($api_action eq "tag") {    # api/v3/tag/[type]/[tagid]
```

File: lib/ProductOpener/Routing.pm [hunk 4]
```diff
@@ -63,137 +64,561 @@ use Log::Any qw($log);
...
+	my @product_route = (map {[$_, \&product_route]} values %{$tag_type_singular{products}});
+	# all translations for route 'en:product' (e.g. fr:produit, es:producto ...)
+	my @lc_product_route
+		= (map {["$_:$tag_type_singular{products}{$_}", \&product_route]} keys %{$tag_type_singular{products}});
+	# text route : index, index-pro, ...
+	my @text_route = (map {[$_, \&text_route]} keys %texts);
+	# Renamed text : en/nova-groups-for-food-processing -> nova, ...
+	my @redirect_text_route = ();
+	if (defined $options{redirect_texts}) {
+		# we use a custome regex to exactly match "en/nova-groups-for-food-processing"
```
Comment (stephanegigandet) [inline lib/ProductOpener/Routing.pm:127]:
```suggestion
		# we use a custom regex to exactly match "en/nova-groups-for-food-processing"
```

File: lib/ProductOpener/Routing.pm [hunk 5]
```diff
@@ -63,137 +64,561 @@ use Log::Any qw($log);
...
+		redirect_to_url($request_ref, 302, $request_ref->{redirect});
+	}
+	else {
+		$request_ref->{status_code} = 404;
+		$request_ref->{error_message} = lang("error_invalid_address");
+	}
+	return 1;
+}
+
+sub handle_other_tag_types_in_route($request_ref, @components) {
```
Comment (stephanegigandet) [inline lib/ProductOpener/Routing.pm:447]: Maybe this function could be renamed to facets_route()

File: lib/ProductOpener/Routing.pm [hunk 6]
```diff
@@ -110,6 +110,8 @@ sub load_routes() {
 		['', \&index_route],
 		['^(?<page>\d+)$', \&index_route, {regex => 1}],
 		['org/[orgid]/', \&org_route],
+		# Known tag type? Catch all if no route matched
```
Comment (stephanegigandet) [inline lib/ProductOpener/Routing.pm:113]: Great

**Review by stephanegigandet — APPROVED** [review]
Looks good to me, thank you for all the refactor work!

**Pull request merged**